### PR TITLE
initial ijson commit

### PR DESCRIPTION
**Pull request by gkorland**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -11,10 +11,11 @@ name = "rejson"
 [dependencies]
 log = "0.4"
 bson = "0.14"
+ijson = "0.1.3"
 serde_json = "1.0"
 serde = "1.0"
 libc = "0.2"
-jsonpath_lib = { git = "https://github.com/RedisJSON/jsonpath.git", branch = "generic_json_path" }
+jsonpath_lib = { git = "https://github.com/RedisJSON/jsonpath.git", branch = "gkorland-ijson" }
```
Comment (MeirShpilraien) [inline Cargo.toml:18]: Change back once merged on jsonpath repository

**Comment by MeirShpilraien** [comment]
- [ ] Fix inaccurate tests
- [ ] Allow json back-end (serde_json:Value or IValue) to be configurable on startup

File: src/ivalue_manager.rs [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+                    let v = if !docs.is_empty() {
+                        docs.iter().next().map_or_else(
+                            || IValue::NULL,
+                            |(_, b)| {
+                                let v: serde_json::Value = b.clone().into();
+                                let mut out = serde_json::Serializer::new(Vec::new());
+                                v.serialize(&mut out).unwrap();
+                                self.from_str(
+                                    &String::from_utf8(out.into_inner()).unwrap(),
+                                    Format::JSON,
```
Comment (oshadmi) [inline src/ivalue_manager.rs:513]: Not for now (there are missing implementations and tests for BSON), but maybe could be done without passing through JSON string?

**Review by oshadmi — APPROVED** [review]
Looks great! Just small non-blocking comments

File: src/lib.rs [hunk 3]
```diff
@@ -462,7 +491,12 @@ fn dummy_init(_ctx: &Context, _args: &Vec<RedisString>) -> Status {
 redis_json_module_create! {
     data_types: [REDIS_JSON_TYPE],
     pre_command_function: pre_command,
-    get_manage: Some(manager::RedisJsonKeyManager{phantom:PhantomData}),
+    get_manage: {
+        match MANAGER {
+            ManagerType::IValue => Some(ivalue_manager::RedisIValueJsonKeyManager{phantom:PhantomData}),
+            _ => None,
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/lib.rs:497]: No, because then you have a match that return different types

File: src/ivalue_manager.rs (hunk 2 above)
Comment (MeirShpilraien) [reply src/ivalue_manager.rs:513]: We need to add support for this on ijson repository

File: src/ivalue_manager.rs [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+        &mut self,
+        path: Vec<String>,
+        key: &str,
+        mut v: IValue,
+    ) -> Result<bool, RedisError> {
+        let mut updated = false;
+        if path.is_empty() {
+            // update the root
+            let root = self.get_value().unwrap().unwrap();
+            let val = if root.is_object() {
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/ivalue_manager.rs]: Fixed

**Review by oshadmi — APPROVED** [review]

**Comment by oshadmi** [comment]
Fix #569

**Review by oshadmi — APPROVED** [review]
👍🏼

**Comment by filipecosta90** [comment]
@gkorland @MeirShpilraien @oshadmi , adding a bit of context of the small regressions this PR will add to the achievable ops/sec ( at the benefit of large memory savings ). 
This table has already been discussed deeply with @oshadmi and the team. The bellow tables are only focusing on achievable ops/sec and don't focus on the benefits of memory savings. ( lets say this focus only on the bad part :) )

**regressions/potential regressions only:**
```
# Comparison between master and gkorland-ijson for metric: Tests.Overall.rps. Time Period from 7 days ago to now
|                 Test Case                 |Baseline (median obs. +- std.dev)|Comparison (median obs. +- std.dev)|% change (higher-better)|           Note           |
|-------------------------------------------|---------------------------------|-----------------------------------|------------------------|--------------------------|
|json_get_[web-app].servlet_json-parser-0000| 83670.800 +- 4.2%               | 76478.290 +- 4.2%                 |-8.6%                   |REGRESSION                |
|json_get_array_of_docs_pass_100_json       | 134220.980 +- 5.4%              | 126385.935 +- 5.6%                |-5.8%                   |waterline=5.6%. REGRESSION|
|json_get_fulldoc_json-parser-0000          | 91112.125 +- 4.6%               | 79830.445 +- 0.7%                 |-12.4%                  |REGRESSION                |
|json_get_fulldoc_jsonsl-1                  | 103082.160 +- 3.5%              | 94332.820 +- 1.2%                 |-8.5%                   |REGRESSION                |
|json_get_fulldoc_jsonsl-yahoo2_json        | 28563.270 +- 4.8%               | 25967.365 +- 3.0%                 |-9.1%                   |REGRESSION                |
|json_get_fulldoc_jsonsl-yelp_json          | 14483.940 +- 4.0%               | 13019.615 +- 1.7%                 |-10.1%                  |REGRESSION                |
|json_get_key_empty                         | 159994.890 +- 4.6%              | 146046.905 +- 7.6%                |-8.7%                   |waterline=7.6%. REGRESSION|
|json_get_sub_doc_pass_100_json             | 140837.140 +- 5.3%              | 126174.675 +- 6.5%                |-10.4%                  |waterline=6.5%. REGRESSION|
|json_set_num_0                             | 153836.690 +- 5.0%              | 144919.140 +- 4.4%                |-5.8%                   |waterline=5.0%. REGRESSION|
```

**full data:**

```
2021-12-09 19:32:57,027 INFO Using deployment_type=oss-standalone and deployment_name=oss-standalone for the analysis
2021-12-09 19:32:57,028 INFO Using a time-delta from 7 days ago to now
2021-12-09 19:32:57,123 WARNING Based on test-cases set (key=ci.benchmarks.redislabs/circleci/RedisJSON/RedisJSON:testcases) we have 43 distinct benchmarks. 
2021-12-09 19:33:05,519 INFO Printing differential analysis between branches
# Comparison between master and gkorland-ijson for metric: Tests.Overall.rps. Time Period from 7 days ago to now
|                           Test Case                            |Baseline (median obs. +- std.dev)|Comparison (median obs. +- std.dev)|% change (higher-better)|                Note                 |
|----------------------------------------------------------------|---------------------------------|-----------------------------------|------------------------|-------------------------------------|
|json_arrappend_geojson                                          | 60141.085 +- 4.2%               | 57546.995 +- 3.6%                 |-4.3%                   |potential REGRESSION                 |
|json_get_ResultSet.totalResultsAvailable_jsonsl-yahoo2_json     | 140833.170 +- 4.3%              | 130518.540 +- 9.4%                |-7.3%                   |waterline=9.4%. potential REGRESSION |
|json_get_[0]_jsonsl-1                                           | 132443.310 +- 5.4%              | 136549.155 +- 5.0%                |3.1%                    |waterline=5.4%. potential IMPROVEMENT|
|json_get_[7]_jsonsl-1                                           | 144919.140 +- 4.1%              | 146513.525 +- 0.9%                |1.1%                    |potential IMPROVEMENT                |
|json_get_[8].zero_jsonsl-1                                      | 134220.980 +- 4.6%              | 132894.710 +- 3.4%                |-1.0%                   |potential REGRESSION                 |
|json_get_[web-app].servlet[0][servlet-name]_json-parser-0000    | 119415.315 +- 5.4%              | 117633.210 +- 3.1%                |-1.5%                   |waterline=5.4%. potential REGRESSION |
|json_get_[web-app].servlet[0]_json-parser-0000                  | 90489.550 +- 4.2%               | 88886.430 +- 3.9%                 |-1.8%                   |potential REGRESSION                 |
|json_get_[web-app].servlet_json-parser-0000                     | 83670.800 +- 4.2%               | 76478.290 +- 4.2%                 |-8.6%                   |REGRESSION                           |
|json_get_array_of_docs[1]_pass_100_json                         | 131568.560 +- 5.4%              | 125798.115 +- 1.7%                |-4.4%                   |waterline=5.4%. potential REGRESSION |
|json_get_array_of_docs[1]sclr_pass_100_json                     | 133326.220 +- 3.5%              | 130717.710 +- 1.5%                |-2.0%                   |potential REGRESSION                 |
|json_get_array_of_docs_pass_100_json                            | 134220.980 +- 5.4%              | 126385.935 +- 5.6%                |-5.8%                   |waterline=5.6%. REGRESSION           |
|json_get_fulldoc_json-parser-0000                               | 91112.125 +- 4.6%               | 79830.445 +- 0.7%                 |-12.4%                  |REGRESSION                           |
|json_get_fulldoc_jsonsl-1                                       | 103082.160 +- 3.5%              | 94332.820 +- 1.2%                 |-8.5%                   |REGRESSION                           |
|json_get_fulldoc_jsonsl-yahoo2_json                             | 28563.270 +- 4.8%               | 25967.365 +- 3.0%                 |-9.1%                   |REGRESSION                           |
|json_get_fulldoc_jsonsl-yelp_json                               | 14483.940 +- 4.0%               | 13019.615 +- 1.7%                 |-10.1%                  |REGRESSION                           |
|json_get_fulldoc_pass_100_json                                  | 135127.830 +- 4.4%              | 132017.395 +- 5.9%                |-2.3%                   |waterline=5.9%. potential REGRESSION |
|json_get_key_empty                                              | 159994.890 +- 4.6%              | 146046.905 +- 7.6%                |-8.7%                   |waterline=7.6%. REGRESSION           |
|json_get_message.code_jsonsl-yelp_json                          | 136978.800 +- 5.8%              | 136640.225 +- 4.6%                |-0.2%                   |waterline=5.8%. potential REGRESSION |
|json_get_sclr_pass_100_json                                     | 142848.980 +- 6.6%              | 149253.145 +- 5.8%                |4.5%                    |waterline=6.6%. potential IMPROVEMENT|
|json_get_sub_doc.sclr_pass_100_json                             | 136047.020 +- 4.8%              | 137947.820 +- 2.8%                |1.4%                    |potential IMPROVEMENT                |
|json_get_sub_doc_pass_100_json                                  | 140837.140 +- 5.3%              | 126174.675 +- 6.5%                |-10.4%                  |waterline=6.5%. REGRESSION           |
|json_numincrby_num_1                                            | 138881.170 +- 4.2%              | 139857.170 +- 3.1%                |0.7%                    |potential IMPROVEMENT                |
|json_nummultby_num_2                                            | 136043.330 +- 5.8%              | 140837.140 +- 2.7%                |3.5%                    |waterline=5.8%. potential IMPROVEMENT|
|json_set_ResultSet.totalResultsAvailable_1_jsonsl-yahoo2_json   | 118331.980 +- 3.9%              | 109944.655 +- 7.5%                |-7.1%                   |waterline=7.5%. potential REGRESSION |
|json_set_[0]foo_jsonsl-1                                        | 121203.310 +- 5.6%              | 124993.965 +- 5.2%                |3.1%                    |waterline=5.6%. potential IMPROVEMENT|
|json_set_[web-app].servlet[0][servlet-name]_bar_json-parser-0000| 95229.020 +- 4.0%               | 91604.275 +- 5.6%                 |-3.8%                   |waterline=5.6%. potential REGRESSION |
|json_set_fulldoc_api_replies_q1_google_autocomplete             | 41949.015 +- 28.8% UNSTABLE     | 42005.690 +- 9.9%                 |0.1%                    |UNSTABLE (very high variance)        |
|json_set_fulldoc_api_replies_q2_gmaps_areatraffic               | 76697.295 +- 29.7% UNSTABLE     | 73250.950 +- 18.9% UNSTABLE       |-4.5%                   |UNSTABLE (very high variance)        |
|json_set_fulldoc_api_replies_q3_gmaps_passiveassist             | 39545.540 +- 28.8% UNSTABLE     | 37405.970 +- 11.9% UNSTABLE       |-5.4%                   |UNSTABLE (very high variance)        |
|json_set_fulldoc_api_replies_q4_gmaps_assist                    | 96183.130 +- 28.5% UNSTABLE     | 97316.020 +- 6.8%                 |1.2%                    |UNSTABLE (very high variance)        |
|json_set_fulldoc_pass-json-parser-0000                          | 32309.590 +- 27.2% UNSTABLE     | 31236.920 +- 5.7%                 |-3.3%                   |UNSTABLE (very high variance)        |
|json_set_fulldoc_pass_100_json                                  | 90080.350 +- 9.2%               | 85151.740 +- 8.9%                 |-5.5%                   |waterline=9.2%. potential REGRESSION |
|json_set_fulldoc_yahoo2                                         | 7747.390 +- 13.7% UNSTABLE      | 7939.600 +- 2.6%                  |2.5%                    |UNSTABLE (very high variance)        |
|json_set_key_empty                                              | 157470.390 +- 4.5%              | 151514.675 +- 7.0%                |-3.8%                   |waterline=7.0%. potential REGRESSION |
|json_set_message.code_1_jsonsl-yelp_json                        | 117638.760 +- 6.2%              | 116610.910 +- 0.3%                |-0.9%                   |waterline=6.2%. potential REGRESSION |
|json_set_num_0                                                  | 153836.690 +- 5.0%              | 144919.140 +- 4.4%                |-5.8%                   |waterline=5.0%. REGRESSION           |
|json_set_sclr_1_pass_100_json                                   | 128195.270 +- 5.0%              | 122856.800 +- 5.4%                |-4.2%                   |waterline=5.4%. potential REGRESSION |
|json_set_sclr_pass_100_json                                     | 121203.310 +- 6.5%              | 116634.705 +- 6.8%                |-3.8%                   |waterline=6.8%. potential REGRESSION |
|json_vs_hashes_hset_key_simple                                  | 164600.920 +- 5.2%              | 161799.155 +- 4.7%                |-1.7%                   |waterline=5.2%. potential REGRESSION |
|json_vs_hashes_json.set_key_simple                              | 143876.610 +- 4.7%              | 144384.900 +- 5.0%                |0.4%                    |potential IMPROVEMENT                |
2021-12-09 19:33:05,622 INFO Detected a total of 25 stable tests between versions.
2021-12-09 19:33:05,622 WARNING Detected a total of 6 highly unstable benchmarks.
2021-12-09 19:33:05,622 WARNING Detected a total of 9 regressions bellow the regression water line.
```

**Pull request merged**